### DISCUSSION TITLE
Passing a cached computed property that contains an Eloquent collection causes the query to run again on update

### DISCUSSION THREAD

**Discussion by afonic** [livewire/livewire#8202] [1. Help] [answered]
@joshhanley here it is!

I'll admit this is a bit weird. I have a nested components setup. In the parent `Checkout` component I use a computed property to hold some Eloquent data that won't be updated like so:

```php
#[Computed(persist: true)]
public function options(): Collection
{
    return Option::entry($reservation->item_id)
        ->where('published', true)
        ->with('values')
        ->get();
}
```

Now in a child `CheckoutOptions` component I pass the `$options`:
```blade
<livewire:checkout-options wire:model.live="enabledOptions" :options="$this->options" />
```

in `CheckoutOptions` I accept this property and set it to locked:
```php
#[Locked]
public Collection $options;
```

Finally I loop into the `CheckoutOptions` view to show said options and display yet another Livewire component:
```blade
@foreach ($options as $id => $option)
    <livewire:option :option="$option" :key="$id" />
@endforeach
```

Everything works as expected, however I noticed that when I update the component (for example an option gets selected), the `Option` model gets called again and fetches the options from the database. I narrowed this down on the `@foreach` loop of `Option` view and Laravel Debugbar says that its the `EloquentCollectionSynth.php` that calls the database. Note that I don't have `legacy_model_binding` enabled as this is a new project. I also tried `#[Computed(cache: true)]` but the result is the same.

I discovered this because I was changing some stuff in the `$options` computed property after fetching the data from the database but when I clicked an option, it got refreshed and back to the way it is in the database. I fixed this by converting the Eloquent collection to an array for the time being.

Thanks for creating Livewire and for the time you'll put into reading about this issue!

PS. After I typing this I realised I didn't try disabling the Debugbar, so I just tried it, but nothing changed.

**Comment by joshhanley** [comment, marked as answer]
@afonic no worries! 🙂 So what the problem is, is the fact that you are passing the `$options` into the `CheckoutOptions` nested component and assigning that to a public property. That is what is causing the `EloquentCollectionSynth` to be handling it.

In your `Checkout` component it will be fine, as the options are computed.

The things that are likely to be missing when the `EloquentCollectionSynth` rehydrates the collection is any custom query stuff (like selects, joins, etc) and relations, which will likely be the weirdness that you have experienced. You can read more about the limitations here https://livewire.laravel.com/docs/properties#eloquent-constraints-arent-preserved-between-requests

You might be wondering how best to handle it then? And it just depends on what you are trying to achieve. Ideally you wouldn't pass eloquent models/ eloquent collections down to child components like that and assign them to public properties (although sometimes it is necessary). At the moment though we don't have any way to pass data into a property and sort of "prefill" a computed property which makes it more challenging.

The question is, what is the purpose/ why do you have 3 different Livewire components" `Checkout`, `CheckoutOptions` and `Option`? Can they be made into Blade components instead?

**Comment by afonic** [comment]
Thanks @joshhanley ! I thought that the problem might be in the way I am handling the `$options` property. 

I'm using a mix and match of Blade and Livewire components in this project:

- The basic `Checkout` component needs to have knowledge of all the options, extras and more because it passes it over to various components (like a payment table) and takes care of the final checkout. 
- The child `CheckoutOptions` component (and a couple more that are nested here) I might be able to rework into a Blade component with some AlpineJS help but it also takes care of validation, i.e. "is this option required" which I would then have to move into `Checkout` making it huge.
- And the `Option` component gets used again in another component 

(it's a fairly complex project - not big but needs to cover a lot of options)

Probably my best option is to work into converting `CheckoutOptions` and `Option` to Blade components so that I don't need to pass around Eloquent collections using Javascript. However every time I do this there comes a time that I say "ah I wish this was a Livewire component so that I can do X", because Livewire has all those goodies hehe.

Thanks again for your time!

**Reply by joshhanley** [reply]
Ah yeah fair enough. Yeah worth a try!

No worries! 🙂

**Reply by afonic** [reply]
Just for the record, I took your advice and with some tinkering I turned them into Blade components. Much faster and everything works fine!

**Reply by joshhanley** [reply]
@afonic nice! Great to hear 😁
